code/WorkInProgress/AzrunStuff.dm

```diff
@@ -57,5 +59,165 @@
+
+			var/reagents_present = length(T.required_reagents)
+			for(var/R as() in T.required_reagents)
+				if(generator.circ1.reagents.get_reagent_amount(R) + generator.circ2.reagents.get_reagent_amount(R) >= T.required_reagents[R])
+				else
+					reagents_present = FALSE
+					break
+
+			if(reagents_present)
+				SPAWN_DBG(0)
+					if(generator.active_form)
+						generator.active_form.on_revert()
+					generator.active_form = new T.type
+					generator.active_form.on_transform(generator)
+				return
+
+	/// Transform when a matsci semiconductor is inserted and the material differs the material
+	/// from the TEG.  Transformation requires the semiconductor fully back in place and energy
+	/// is present to activate NANITES!
+	proc/check_material_transformation()
+		if(!generator.active_form || generator.active_form.type == /datum/teg_transformation/matsci)
+			if(generator.semiconductor?.material && ((src.generator.semiconductor.material.mat_id != src.generator.material?.mat_id) || !src.generator.material))
+				if(src.generator.semiconductor_state == 0 && src.generator.powered())
+					SPAWN_DBG(1.5 SECONDS)
+						src.generator.use_power(500 WATTS)
+						elecflash(src.generator)
+						src.generator.visible_message("<span class='alert'>[src.generator] is suddenly engulfed in a swarm of nanites!</span>")
+						var/nanite_overlay = image('icons/misc/critter.dmi', "nanites")
+						generator.UpdateOverlays(nanite_overlay,"transform")
+						generator.circ1.UpdateOverlays(nanite_overlay,"transform")
+						generator.circ2.UpdateOverlays(nanite_overlay,"transform")
+						sleep(rand(1.5 SECONDS, 2.5 SECONDS))
+						if(generator.active_form)
+							generator.active_form.on_revert()
+						generator.active_form = new /datum/teg_transformation/matsci
+						generator.active_form.mat_id = generator.semiconductor.material.mat_id
+						generator.active_form.on_transform(generator)
+						sleep(rand(1.5 SECONDS, 2.5 SECONDS))
+						src.generator.visible_message("<span class='alert'>The swarm of nanites disappears back into \the [src.generator].</span>")
+						generator.UpdateOverlays(null,"transform")
+						generator.circ1.UpdateOverlays(null,"transform")
+						generator.circ2.UpdateOverlays(null,"transform")
+
+
+ABSTRACT_TYPE(/datum/teg_transformation)
+/** Thermo-Electric Generator Transformations
+	These are various forms the Thermo-Electric Generator can take. They can
+	be achieved by:
+		* a reagent mixture similar to chems
+		* triggered directly via another condition
+  */
+datum/teg_transformation
+	var/name = null
+	/// material id to apply
+	var/mat_id
+	/// associated list of reagent ids and amounts to cause transformation
+	var/list/required_reagents
+	/// ref to TEG
+	var/obj/machinery/power/generatorTemp/teg
+	/// Automatic transformation checks until a seperate criteria is achieved
+	var/skip_transformation_checks = FALSE
+
+	disposing()
+		. = ..()
+		teg = null
```

Comment:
in disposing it generally makes sense to call parent *after* your own stuff. (Think of it as a closing parenthesis to the New's opening parenthesis where it's the other way around.)
```suggestion
		teg = null
		. = ..()
```

Reply:
(Doesn't really change anything in this particular case but it's good practice to stick to this. Parent's disposing can clean up stuff you might have wanted to override by cleaning it up yourself, especially with `contents`.)

Reply:
Hmm like a stack you say. 👍 But this stops me from possibly disposing something I'm going to reference later. 😈

Reply:
I swear your second comment magically appeared. 🧙

Reply:
Updated `/datum/teg_transformation_mngr` as well to match.